jormungandr/src/network/protocols/bootstrap.rs

```diff
@@ -0,0 +1,141 @@
+use super::super::{grpc, BlockConfig};
+use crate::blockcfg::{Block, HeaderHash};
+use crate::blockchain::protocols::{Blockchain, Error as BlockchainError, PreCheckedHeader, Ref};
+use crate::settings::start::network::Peer;
+use chain_core::property::HasHeader;
+use network_core::client::{block::BlockService, Client as _};
+use network_core::error::Error as NetworkError;
+use network_grpc::client::Connection;
+use slog::Logger;
+use tokio::prelude::*;
+use tokio::runtime::current_thread;
+
+use std::error;
+use std::fmt::{self, Debug, Display};
+
+type ConnectError = network_grpc::client::ConnectError<std::io::Error>;
+
+#[derive(Debug)]
+pub enum Error {
+    Connect(ConnectError),
+    ClientNotReady(NetworkError),
+    PullRequestFailed(NetworkError),
+    PullStreamFailed(NetworkError),
+    HeaderCheckFailed(BlockchainError),
+    BlockAlreadyPresent(HeaderHash),
+    BlockMissingParent(HeaderHash),
+    ApplyBlockFailed(BlockchainError),
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+        match self {
+            Connect(_) => write!(f, "failed to connect to bootstrap peer"),
+            ClientNotReady(_) => write!(f, "connection broken"),
+            PullRequestFailed(_) => write!(f, "bootstrap pull request failed"),
+            PullStreamFailed(_) => write!(f, "bootstrap pull stream failed"),
+            HeaderCheckFailed(_) => write!(f, "block header check failed"),
+            BlockAlreadyPresent(hash) => write!(f, "received block {} is already present", hash),
+            BlockMissingParent(hash) => write!(
+                f,
+                "received block {} is not connected to the block chain",
+                hash
+            ),
+            ApplyBlockFailed(_) => write!(f, "failed to apply block to the blockchain"),
+        }
+    }
+}
+
+impl error::Error for Error {
+    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
+        use self::Error::*;
+        match self {
+            Connect(e) => Some(e),
+            ClientNotReady(e) => Some(e),
+            PullRequestFailed(e) => Some(e),
+            PullStreamFailed(e) => Some(e),
+            HeaderCheckFailed(e) => Some(e),
+            BlockAlreadyPresent(_) => None,
+            BlockMissingParent(_) => None,
+            ApplyBlockFailed(e) => Some(e),
+        }
+    }
+}
+
+pub fn bootstrap_from_peer(
+    peer: Peer,
+    blockchain: Blockchain,
+    tip: Ref,
+    logger: &Logger,
+) -> Result<Ref, Error> {
+    info!(logger, "connecting to bootstrap peer {}", peer.connection);
+    let bootstrap = grpc::connect(peer.address(), None)
+        .map_err(Error::Connect)
+        .and_then(|client: Connection<BlockConfig>| client.ready().map_err(Error::ClientNotReady))
```

Comment:
you might need to call `Branch.get_ref` (which is a future) to get the current `Ref` from the `Branch`